packages/compiler/src/render3/r3_deferred_triggers.ts

```diff
@@ -0,0 +1,344 @@
+    // Eventually we could expose this information on the token directly.
+    return this.expression.slice(this.start + this.token().index, this.start + this.token().end);
+  }
+
+  private error(token: Token, message: string): void {
+    const newStart = this.span.start.moveBy(this.start + token.index);
+    const newEnd = newStart.moveBy(token.end - token.index);
+    this.errors.push(new ParseError(new ParseSourceSpan(newStart, newEnd), message));
+  }
+
+  private unexpectedToken(token: Token) {
+    this.error(token, `Unexpected token "${token}"`);
+  }
+}
+
+function createIdleTrigger(
+    parameters: string[], sourceSpan: ParseSourceSpan): t.IdleDeferredTrigger {
+  if (parameters.length > 0) {
+    throw new Error(`"${OnTriggerType.IDLE}" trigger cannot have parameters`);
+  }
+
+  return new t.IdleDeferredTrigger(sourceSpan);
+}
+
+function createTimerTrigger(parameters: string[], sourceSpan: ParseSourceSpan) {
+  if (parameters.length !== 1) {
+    throw new Error(`"${OnTriggerType.TIMER}" trigger must have exactly one parameter`);
+  }
+
+  const delay = parseDeferredTime(parameters[0]);
+
+  if (delay === null) {
+    throw new Error(`Could not parse time value of trigger "${OnTriggerType.TIMER}"`);
+  }
+
+  return new t.TimerDeferredTrigger(delay, sourceSpan);
+}
+
+function createInteractionTrigger(
+    parameters: string[], sourceSpan: ParseSourceSpan): t.InteractionDeferredTrigger {
+  if (parameters.length > 1) {
+    throw new Error(`"${OnTriggerType.INTERACTION}" trigger can only have zero or one parameters`);
+  }
+
+  return new t.InteractionDeferredTrigger(parameters[0] ?? null, sourceSpan);
+}
+
+function createImmediateTrigger(
+    parameters: string[], sourceSpan: ParseSourceSpan): t.ImmediateDeferredTrigger {
+  if (parameters.length > 0) {
+    throw new Error(`"${OnTriggerType.IMMEDIATE}" trigger cannot have parameters`);
+  }
+
+  return new t.ImmediateDeferredTrigger(sourceSpan);
+}
+
+function createHoverTrigger(
+    parameters: string[], sourceSpan: ParseSourceSpan): t.HoverDeferredTrigger {
+  if (parameters.length > 0) {
+    throw new Error(`"${OnTriggerType.HOVER}" trigger cannot have parameters`);
+  }
+
+  return new t.HoverDeferredTrigger(sourceSpan);
+}
+
+function createViewportTrigger(
+    parameters: string[], sourceSpan: ParseSourceSpan): t.ViewportDeferredTrigger {
+  // TODO: the RFC has some more potential parameters for `viewport`.
+  if (parameters.length > 1) {
+    throw new Error(`"${OnTriggerType.VIEWPORT}" trigger can only have zero or one parameters`);
+  }
+
+  return new t.ViewportDeferredTrigger(parameters[0] ?? null, sourceSpan);
+}
+
+/** Gets the index within an expression at which the trigger parameters start. */
+export function getTriggerParametersStart(value: string, startPosition = 0): number {
+  let hasFoundSeparator = false;
+
+  for (let i = startPosition; i < value.length; i++) {
+    if (value[i] === ' ' || value[i] === '\n') {
```

Comment:
@crisbeto should we use `\s` here to mimic what we have in other RegExps?
```suggestion
    if (/\s/.test(value[i])) {
```